### ISSUE TITLE
How to modify java:comp/BeanManager to java:comp/env/BeanManager?

### ISSUE THREAD

**Issue by iwangxiaodong** [mybatis/cdi#74]
InitialContext.doLookup("java:comp/BeanManager") in org.mybatis.cdi.CDIUtils.getBeanManager()

Jetty only allows you to bind entries to java:comp/env, so the BeanManager will be available at java:comp/env/BeanManager

How to modify java:comp/BeanManager to java:comp/env/BeanManager?

Thanks!

**Comment by hazendaz** [comment]
Good catch!  This would affect tomcat too as its jndi is read only as well.  So bean manager is going to be in the location you noted above.  I'd say drop back to 1.0.0 release for time being.

**Comment by iwangxiaodong** [comment]
I copied CDIUtils.java and modified it as follows:

    private static BeanManager getBeanManager() {
        try {
            return InitialContext.doLookup("java:comp/BeanManager");
        } catch (NamingException e) {
            try {
                return InitialContext.doLookup("java:comp/env/BeanManager");
            } catch (NamingException ex) {
                throw new RuntimeException(ex);
            }
        }
    }

It works fine！

**Comment by hazendaz** [comment]
@iwangxiaodong Can you kindly send a PR for this.  Thanks.  I've used this exact same thing for JNDI many times in the past.  While not the most efficient, given it hits multiple container types it's probably the easiest.

**Comment by mnesarco** [comment]
@iwangxiaodong @hazendaz It would be nice if we can cache at least the jndi name. but maybe it requires some synchronization. I think throwing exceptions and catching them every time this method is called is a lot of overhead. We have to decide what is less bad: synchronization vs try/catch. Or maybe we can statically do a lookup when the class is loaded and save the correct jndi name, so we will not require sync or try/catch inside the method.

**Comment by mnesarco** [comment]
@iwangxiaodong @hazendaz What do you think about something like this:

```
  private static final String DEFAULT_JNDI_NAME = "java:comp/BeanManager";
  private static final AtomicBoolean INITIALIZED = new AtomicBoolean(false);
  private static String jndiName = DEFAULT_JNDI_NAME; 

  private static void init() {
    if (!INITIALIZED.getAndSet(true)) {
      synchronized (DEFAULT_JNDI_NAME) {
        try {
          if (InitialContext.doLookup("java:comp/env/BeanManager") != null) {
            jndiName = "java:comp/env/BeanManager";
          }
        } catch (NamingException e) {
          // Fallback to default, do nothing
        }
      }
    }
  }

  private static BeanManager getBeanManager() {
    init();
    try {
      return InitialContext.doLookup(jndiName);
    } catch (NamingException e) {
      throw new RuntimeException(e);
    }
  }
```

**Comment by mnesarco** [comment]
@iwangxiaodong @hazendaz Or even better:

```
public final class CDIUtils {

  private static class JNDI {

    private static final String DEFAULT_JNDI_NAME = "java:comp/BeanManager";
    static final String NAME;

    static {
      String jndiName = DEFAULT_JNDI_NAME;
      try {
        if (InitialContext.doLookup("java:comp/env/BeanManager") != null) {
          jndiName = "java:comp/env/BeanManager";
        }
      } catch (NamingException e) {
        // Fallback to default, do nothing
      }
      NAME = jndiName;
    }

  }

  private static BeanManager getBeanManager() {
    try {
      return InitialContext.doLookup(JNDI.NAME);
    } catch (NamingException e) {
      throw new RuntimeException(e);
    }
  }


```

**Comment by iwangxiaodong** [comment]
@mnesarco @hazendaz This code is very good performance!

**Comment by mnesarco** [comment]
@iwangxiaodong Thanks for the catch. I will submit the patch just now.

**Comment by iwangxiaodong** [comment]
@mnesarco Yeah！ Nice！

**Comment by hazendaz** [comment]
Great!  I can push a release tonight.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by hazendaz** [comment]
@iwangxiaodong @mnesarco I have released version 1.0.2 with this change.  It should show in central in about 2 hours but is otherwise available for download now.

**Comment by mnesarco** [comment]
Thank you very much Jeremy! Nice work!

El 13/10/2017 9:15 PM, "Jeremy Landis" <notifications@github.com> escribió:

@iwangxiaodong <https://github.com/iwangxiaodong> @mnesarco
<https://github.com/mnesarco> I have released version 1.0.2 with this
change. It should show in central in about 2 hours but is otherwise
available for download now.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/mybatis/cdi/issues/74#issuecomment-336602811>, or mute
the thread

**Comment by iwangxiaodong** [comment]
@hazendaz 

compile 'org.mybatis:mybatis-cdi:1.0.2'

I tested OK!
